net: ethernet: Set the ptype by the caller in send

jukkar opened the pull request:
Instead of setting the upper protocol type in ethernet_send() by checking the protocol type bits, use the ptype that is already set by the caller. This allows new protocol types to be supported and makes the system extensible

jukkar committed:
net: ethernet: Set the ptype by the caller in send

Instead of setting the upper protocol type in ethernet_send()
by checking the protocol type bits, use the ptype that is already
set by the caller. This allows new protocol types to be supported
and makes the system extensible.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

clamattia reviewed (commented):
Nice, thanks jukkar.

Maybe in a next step custom bit-flags like `net_pkt_set_lldp` and similar could be removed.

rlubos commented on subsys/net/l2/ethernet/arp.c:
```diff
@@ -270,6 +270,9 @@ static inline struct net_pkt *arp_prepare(struct net_if *iface,
 			return NULL;
 		}
 
+		net_pkt_set_ll_proto_type(pkt, NET_ETH_PTYPE_ARP);
```
In ARP context, `arp_gratuitous_send()` needs an update too.

jukkar commented:
> Maybe in a next step custom bit-flags like `net_pkt_set_lldp` and similar could be removed.

Yes, absolutely. There needs to be a cleanup before this is ready for merge. This is currently a draft so people could try it without it getting merged accidentally.

jukkar committed:
tests: net: arp: Use documentation IP address space

Use the 192.0.2/24 address space reserved for documentation.
No functional changes by this commit.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

clamattia reviewed (approved):
LGTM, thank you for adjusting. As long as it is possible to use `ethernet_send` with some custom `ptype` set, I am happy. I do not understand some of the remaining special handling in said function.

jukkar commented:
While testing things, I noticed some issues with this. For example ping does not work to the device. If I am not able to figure out the culprit soon, I will send a revert for this.

jukkar commented:
The ARP issue I encountered is fixed by #84158
